### ISSUE TITLE
Dockerize

### ISSUE THREAD

**Issue by Pastukhovar** [synesthesiam/homeassistant-satellite#5]
Hi!
What about to run in inside container?

**Comment by chatziko** [comment]
Until someone creates an image, here's a sample Dockerfile to get started. 

```
FROM python:3.11-slim-bookworm AS base-image

RUN apt-get update \
	&& apt-get install --no-install-recommends -y \
        alsa-utils \
    # Clean-up
	&& apt-get clean \
	&& rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/* ~/.cache


#################################################


FROM base-image AS compile-image

RUN apt-get update \
	&& apt-get install --no-install-recommends -y \
        unzip

ADD https://github.com/synesthesiam/homeassistant-satellite/archive/refs/heads/master.zip /tmp
RUN cd /tmp && unzip *.zip && rm *.zip
RUN cd /tmp && mv homeassistant-satellite-* homeassistant-satellite

RUN python3 -m venv /venv

RUN /venv/bin/pip3 --no-cache-dir install \
        /tmp/homeassistant-satellite[silerovad,webrtc]

# copy the sounds
RUN cp -r /tmp/homeassistant-satellite/sounds /venv


# ##############################################3


FROM base-image AS build-image

COPY --from=compile-image /venv /venv

# Copy statically build ffmpeg, much smaller!
COPY --from=mwader/static-ffmpeg:6.0-1 /ffmpeg /usr/local/bin/

CMD [ "/venv/bin/python3", "-m", "homeassistant_satellite", "--host=...", "--token=..." ]
```

**Comment by NonaSuomy** [comment]
Do you have a docker compose to go along with it and how to connect the microphone/speaker in/output ?

**Comment by chatziko** [comment]
I uploaded a fully working solution (at least in my machine) in [this branch](https://github.com/chatziko/homeassistant-satellite/tree/docker). Give it a try, just modify `entrypoint.sh` and run `docker compose up` in the `docker` dir.

**Comment by NonaSuomy** [comment]
Note it says it requires the "add-on" I can't use add-ons in my setup as I run HA in a docker but I do have wyoming-openwakeword running in a docker as well. Will this be ok?
https://github.com/chatziko/homeassistant-satellite/blob/93d319f731678f0c53e4adb11eca0b6b110f859b/README.md?plain=1#L5

**Comment by chatziko** [comment]
Yes, the README is not 100% accurate, wyoming-openwakeword in docker should be perfectly fine.

**Comment by NonaSuomy** [comment]
I'm not sure if this is because of some settings like ignore folder or something but when you git clone your repo it doesn't grab the docker folder.

**Comment by NonaSuomy** [comment]
--awake-sound="/venv/sounds/awake.wav"
--done-sound="/venv/sounds/done.wav" <== This seems to be missing from your config.

**Comment by Fraddles** [comment]
> I uploaded a fully working solution (at least in my machine) in [this branch](https://github.com/chatziko/homeassistant-satellite/tree/docker). Give it a try, just modify `entrypoint.sh` and run `docker compose up` in the `docker` dir.

Works like a charm in Docker on RPi 4 with Poly USB speakerphone.

Running HA, Whisper, Piper, and Openwakeword in plain Docker on an old laptop.  So can confirm the 'add-on' is not required.

Also works with HA using HTTPS if you add the `--https` option to the `entrypoint.sh` file.

Cheers.

**Comment by NonaSuomy** [comment]
Any idea how to select the audio interfaces? I can't seem to get them to respond properly.

aplay test

This test worked for plughw:1,0
```bash
root@7744408c9df3:/# aplay -D plughw:1,0 /usr/share/sounds/alsa/Front_Right.wav
Playing WAVE '/usr/share/sounds/alsa/Front_Right.wav' : Signed 16 bit Little Endian, Rate 48000 Hz, Mono
```

Testing without the device selected (Default) didn't work
```bash
root@7744408c9df3:/# aplay /usr/share/sounds/alsa/Front_Right.wav
ALSA lib pcm_dmix.c:999:(snd_pcm_dmix_open) unable to open slave
aplay: main:831: audio open error: No such file or directory
```
I set this 
```
--snd-device="plughw:1,0"
```

For MIC I didn't have any success

```bash
root@7744408c9df3:/# arecord -L
null
    Discard all samples (playback) or generate zero samples (capture)
hw:CARD=PCH,DEV=0
    HDA Intel PCH, ALC3226 Analog
    Direct hardware device without any conversions
plughw:CARD=PCH,DEV=0
    HDA Intel PCH, ALC3226 Analog
    Hardware device with all software conversions
default:CARD=PCH
    HDA Intel PCH, ALC3226 Analog
    Default Audio Device
sysdefault:CARD=PCH
    HDA Intel PCH, ALC3226 Analog
    Default Audio Device
front:CARD=PCH,DEV=0
    HDA Intel PCH, ALC3226 Analog
    Front output / input
dsnoop:CARD=PCH,DEV=0
    HDA Intel PCH, ALC3226 Analog
    Direct sample snooping device
```

The default test didn't work (worked on the prior docker) but I couldn't get it working with the satellite would record with default arecord but not when I selected the device. 

In this new docker it just errors out no matter what.

```
# arecord test.wav
ALSA lib pcm_dsnoop.c:566:(snd_pcm_dsnoop_open) unable to open slave
arecord: main:831: audio open error: No such file or directory
```

Tried setting ```--mic-device="plughw:1,0"``` 
```
root@7744408c9df3:/# arecord -D plughw:1,0 -d 5 test.wav
arecord: main:831: audio open error: Device or resource busy
```
and tried to leave it commented out as well.

Showing docker container console logs
Test with
```
DEBUG:root:Microphone command: ['arecord', '-D', 'plughw:1,0', '-r', '16000', '-c', '1', '-f', 'S16_LE', '-t', 'raw']
arecord: main:831: audio open error: Device or resource busy
```
Test without
```
DEBUG:root:Microphone command: ['arecord', '-r', '16000', '-c', '1', '-f', 'S16_LE', '-t', 'raw']
ALSA lib pcm_dsnoop.c:566:(snd_pcm_dsnoop_open) unable to open slave
arecord: main:831: audio open error: No such file or directory
```

**Comment by chatziko** [comment]
Are you using a desktop variant of raspberry pi OS? Quite likely pulseaudio is running and keeps the soundcard busy.

For now you can just stop it. Log in as the desktop user and run
```
systemctl --user stop pulseaudio.socket pulseaudio
```
Then run the container.

A version of the satellite with native pulseaudio support will be merged hopefully soon (PR #3), it will make it easier to connect to audio form the container.

**Comment by NonaSuomy** [comment]
I have an Arch Linux server hypervisor based around libvirt + qemu and docker. It was running pulse audio on its i3 desktop. I disabled it `systemctl disable pluseaudio` as you similarly specified for the user then restarted the container. After removing hwplug settings I had specified everything seems to be in working order now. 

The microphone is in a loud-fanned server rack so I'm not sure if it can hear properly just yet but it did arecord loud noise in there and when I aplay'd it back I could hear the loud noise over the speaker and the satellite interfaces no longer complained of in use. 

I'll have to test a recording with voice to see if it can hear anything over the fan noise.

**Comment by databoy2k** [comment]
I've got the Dockerfile, docker-compose.yaml, and entrypoint.sh from @chatziko all set up in my ~/homeassistant-satellite/docker folder, but when I run docker-compose I get a permission denied on /entrypoint.sh. Any thoughts on what went wrong?

**Comment by AnkushMalaker** [comment]
@databoy2k Most probably you need to add execution permission using `chmod +x ./entrypoint.sh`

**Comment by databoy2k** [comment]
> @databoy2k Most probably you need to add execution permission using `chmod +x ./entrypoint.sh`

Thanks. I ended up using the official one on the front rather than the fork. Just haven't had time to figure out why it isn't responding to the hotword, but I'm sure it's a totally different issue now.

**Comment by khalob** [comment]
Is it possible to get this as a docker image in hub.docker.com?

**Comment by galli-miranda** [comment]
Hey, thanks for posting your solution! I have a beginner question... 
once I get the container up and running, do I have to add an entry to my Home Assistant ? I've got it up and running, but I'm either missing a step or have some networking issue to fix. Maybe relevant: My HA instance is not running in network host mode.

**Comment by chatziko** [comment]
In HA you just need to setup the voice assistant (which most likely you have already done), nothing else. And HA needs to be accessible in the url you provide to `--host`. Check the container logs, if the satellite can't connect to HA there should be an error.
